event-store-plans.md

```diff
@@ -0,0 +1,174 @@
+This probably impacts both the `FetchForWriting()` behavior and the inline projection behavior
+
+This absolutely impacts the Wolverine "aggregate handler workflow". There's an opportunity to eliminate some serious inefficiency today
+
+
+## Versioned Projections for Blue/Green deployment
+
+This is really just sticking a "projection version number" on the Projection subclasses that would potentially be used as a suffix on the document
+tables for the projected documents (the document alias). This would enable blue/green deployment of the same projection so that nodes on the previous version can be 
+using the previous revision/definition of the projection while other nodes are running the newer model. 
+
+Obviously there's some serious concern about just how the new projection gets rebuilt in time. I think the "catch up" mode that this document
+introduces as an "Async strategy for FetchForWriting()" is necessary to make this viable
+
+
+## "Poor Man's Load Balancing" of the Async Daemon
+
+See Oskar's post [How to scale out Marten](https://event-driven.io/en/scaling_out_marten/). I'm voting to just bake this work directly into Marten as a 
+first class thing. Do some randomness of the hot/cold behavior for individual projection groups to see if other nodes can take up that work.
+
+I think the async daemon for each tenant database in the case of multi-tenancy through separate databases should be allowed to run on separate nodes. Might take some 
+work with randomness to kind of distribute the load per tenant
+
+By no means will this be a true equivalent of the projection load balancing work planned for "CritterStackPro"
+
+## Dynamic Database per Tenant Multi-Tenancy model
+
+A JasperFx client is sponsoring this work, so it's absolutely in scope. This time there'll need to be a "master" database that has a table
+to track tenants and tenant databases. The async daemon should be able to discover new tenant databases at runtime and try to spin up a new daemon
+for each tenant as a way to de facto discover and activate new tenants.
+
+
+## "First Class Subscription" model
+
+A new model for subscriptions through the async daemon as described by Oskar here: [https://event-driven.io/en/integrating_Marten/](https://event-driven.io/en/integrating_Marten/)
+
+I think this deserves a slightly different abstraction like:
+
+```csharp
+public interface IProjection
+{
+    Task SubscribeAsync(IDocumentOperations operations, IReadOnlyList<IEvent> events,
+        CancellationToken cancellation);
+}
+```
+
+Behind the scenes it's really just just being driven by the daemon running a single page of events at a time, and recording
+the process as it does with projections today.
+
+Now, other issues to consider:
+
+* Is this done as a "hot" subscription that starts wherever the event high water mark is already when a new one is added?
+* Do we support "cold" subscriptions that would cause a new subscription to start from the the beginning at event #1?
```

Comment:
Ideally both?

We have a version of this heavily based on Oskar's example. By default the subscription is cold. If we want to rebuild or make a new one and find that a cold subscription is undesirable, we add a filter at the start of the subscription based on event timestamp and let it catch up skipping older events.

Reply:
I think @oskardudycz is going to agree with that. He's brought that up before.

Reply:
Yup, see: https://github.com/JasperFx/marten/pull/2888/files#r1442806806

See also my write-up on how ESDB subscription works: https://event-driven.io/en/persistent_vs_catch_up_eventstoredb_subscriptions_in_action/. We should enable at least the same set of basic functionalities.